retention: Add bootloader configuration interface

nordicjm opened the pull request:
Adds a bootloader configuration interface which allows for a bootloader (e.g. MCUboot) to set configuration in a shared data area which is then read by the application.

nordicjm committed:
retention: Add bootloader configuration interface

Adds a bootloader configuration interface which allows for a
bootloader (e.g. MCUboot) to set configuration in a shared data
area which is then read by the application.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

Laczen commented:
@nordicjm, it is not needed to use the retention system for this. It can be done without using any driver as shown in https://github.com/zephyrproject-rtos/zephyr/pull/57784.

Laczen reviewed (dismissed):
Can be done without the retention subsystem resulting in smaller code.

henrikbrixandersen commented on subsys/retention/Kconfig:
```diff
@@ -50,6 +50,47 @@ config RETENTION_BOOT_MODE
+
+	  In order to use this, a retention area must be created and set as
+	  the "zephyr,bootloader-config" chosen node via device tree, MCUboot
+	  must be build with the same "zephyr,bootloader-config" DTS node and
```
```suggestion
	  must be built with the same "zephyr,bootloader-config" DTS node and
```

henrikbrixandersen reviewed (dismissed):
Very cool! The Zephyr part looks good (and serves as a good inspiration for exposing other, custom TLVs as well). I am not familiar enough with the MCUboot code to review that PR, sorry.

Laczen commented:
Mcuboot already has a tlv data sharing technique that can be used directly with some memory area (https://github.com/mcu-tools/mcuboot/blob/main/boot/bootutil/src/boot_record.c). It is just needed to point `MCUBOOT_SHARED_DATA_BASE` to the right location.
Instead of introducing a new (zephyr-only) solution the existing solution should be preferred.

nordicjm commented:
> Can be done without the retention subsystem resulting in smaller code.

I decided to use the retention subsystem because it allows validity marking and being positioned absolutely anywhere, both on and off-chip or completely virtually through another core entirely e.g. using something like RPMsg, the world's smallest code size was not the deciding factor behind this decision, it was usability and expandability for the CPUs supported in zephyr today.

> Mcuboot already has a tlv data sharing technique that can be used directly with some memory area

Already addressed above

> Instead of introducing a new (zephyr-only) solution the existing solution should be preferred.

If you don't want a zephyr solution, don't use zephyr, or from your non-zephyr application (which, makes no sense to even use if you're deplying mcuboot through zephyr) point it to the start of the data section after the magic prefix. As a maintainer of multiple systems (MCUboot, MCUmgr, retention subsystem) I decided I wanted to implement it this way, which is a completely valid decision.

Laczen commented:
@nordicjm, I don't quite understand why you selected the retention subsystem for this functionality:

1. Does mcuboot have functionality to share info with running software: yes it does, it is called the shared memory interface,
2. Does zephyr have the means to use this functionality: yes it has, memory-regions can be declared and reserved for this,
3. Can this be used by other bootloaders: yes it can, it is just a memory-region that bootloaders can write to in their own format without any limitations imposed (and with their own validation if needed), a zephyr specific interface for whatever bootloader can easily be constructed,
4. Is it code efficient: yes it is, it requires no code except to allow access to it when using userspace.

Does this cover all possible storage requirements: no it does not, but all devices have ram so it could be selected by any device. There is no need to increase complexity in bootloader code, as every complexity added will result in a failure possibility.

nordicjm commented:
> @nordicjm, I don't quite understand why you selected the retention subsystem for this functionality:
> There is no need to increase complexity in bootloader code

Then don't enable this feature, no one is forcing you to use it and it's not enabled by default.

Laczen commented:
> > @nordicjm, I don't quite understand why you selected the retention subsystem for this functionality:
> > There is no need to increase complexity in bootloader code
> 
> Then don't enable this feature, no one is forcing you to use it and it's not enabled by default.

It is clear to me that you don't want to address my concerns or even consider them. Every line of code added (no matter how well-intentioned) is a failure possibility and should be avoided if possible especially for a bootloader.

nordicjm commented:
> It is clear to me that you don't want to address my concerns or even consider them.

I considered your requests (note: requests because they are not concerns) and my response was posted in https://github.com/zephyrproject-rtos/zephyr/pull/59025#issuecomment-1581989396 and also told you how that you could build mcuboot from zephyr with this feature and use it with non-zephyr applications because the format is the same as it would be for e.g. a mynewt mcuboot build.

nordicjm committed:
tests: boot: Add MCUboot data sharing test

Adds a test for the data sharing retention feature of MCUboot
configuration to an application.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

Laczen reviewed (dismissed):
Blocked to avoid accidental merge.

de-nordic commented on subsys/retention/Kconfig:
```diff
@@ -55,6 +55,46 @@ config RETENTION_BOOT_MODE
+	help
+	  Adds a bootloader configuration sharing system for MCUboot and
+	  applications which allows applications to read the configurion of
+	  MCUboot and the running image. This can be used by applications so
+	  that they know how to e.g. handle firmware updates and place them
```
We will need to modify the `active_image` and `active_slot` functions to be aware of that.

Laczen commented:
The bootloader interface for retention should not enable the retention prefix or crc-type as this would never be changeable. If it would be needed to change the prefix or crc-type the application can not support 2 different prefixes or crc-types.

d3zd3z commented:
> If it would be needed to change the prefix or crc-type the application can not support 2 different prefixes or crc-types.

Why do you say this. It is implemented in code, using a DT format that we define. If we determine that supporting multiple formats is important (and I think it is) we should extend the retention subsystem to support this. The retention subsystem is new, and has not been included in a release, changing its interfaces is fairly easy to do, especially since using it for boot data is one of the two things it was specifically designed for.

Saying that we shouldn't use the retention subsystem because it is lacking a feature, but instead should entirely implement the functionality itself is, frankly, rather silly, especially given that the retention subsystem was mostly created for this specific use case. If there are needed features, let's make sure we discuss them, but they need to be implemented in the right place.

d3zd3z commented:
> Even if by some change it would be possible for the application to support multiple versions (e.g. by creating several retention areas at the same location) the problem is replaced by a problem for mcuboot generation: which one of the several possibilities does it need to choose ?

A given mcuboot would only support one. This would allow an app to work with different versions of mcuboot.

nordicjm committed:
doc: retention: Add blinfo subsystem documentation

Adds some documentation on how to use the new subsystem

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: api: Add blinfo as experimental

Adds blinfo as an experimental API

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.5: Add note on new blinfo subsystem

Adds a note that a new retention module, bootloader information
(blinfo) has been added

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

Laczen commented:
>User presents a very non-standard (likely unique to them) scenario, have explained how it is possible to achieve this using multiple sections in dts then checking which one is valid

The user that presented this was not explained how it is possible to achieve this. If the solution would be solid the user would have no problem with approving it. The user has its doubts about the proposed solution as it is easy to show that providing multiple sections in the dts wil not work in case none of the multiple sections in the dts will pass the validation.
